README.md
# lstash

[![Run tests](https://github.com/kdgm/lstash/actions/workflows/test.yml/badge.svg)](https://github.com/kdgm/lstash/actions/workflows/test.yml)

Lstash is a gem and command line utility to count or grep log messages in a certain time frame from a Logstash Elasticsearch server.

## Installation

Or install it yourself as:

    gem install lstash

## Running lstash from the command line

    $ lstash
    Commands:
      lstash count QUERY     # count number of log messages matching the QUERY
      lstash grep QUERY      # grep log messages from Logstash
      lstash help [COMMAND]  # Describe available commands or one specific command

## The `count` command

    Usage:
      lstash count QUERY

     Description:
       Count log messages matching the QUERY from Logstash and output this count to stdout. QUERY can use Apache Lucene query
       parser syntax.

       Example to count the number of HAProxy log messages in yesterdays month.

       lstash count 'program:haproxy' --from firstday --to today --anchor yesterday

## The `grep` command

    Usage:
      lstash grep QUERY

    Description:
      Grep log messages matching the QUERY from Logstash in ascending timestamp order and output to stdout. QUERY can use Apache    Lucene query parser syntax.

      Example to grep HAProxy log messages from the beginning of this month upto now

      lstash grep 'program:haproxy' --from firstday --to now

## Command line options

    Options:
      -f, [--from=start of time range]                    # date/time, 'now', 'today', 'yesterday', or 'firstday'
      -t, [--to=end of time range]                        # date/time, 'now', 'today', 'yesterday', or 'firstday'
      -a, [--anchor=anchor date/time]                     # used as reference date for firstday
      -e, [--es-url=Elasticsearch endpoint for Logstash]  # or ES_URL environment variable

All times will be relative to the timezone of the machine on which you are running lstash.

## Elasticsearch configuration

By default `lstash` will connnect to Elasticsearch on your localhost as `http://localhost:9200`. To connect
to a different server you can set the `ES_URL` environment variable. URL scheme `http` and port `9200` are default
and may be omitted.

Example

    export ES_URL=log.mydomain.com
    lstash count program:haproxy

Or

    lstash count program:haproxy --es-url log.mydomain.com

## Examples

Count the number of haproxy log messages matching QUERY from Aug 1 at midnight (0:00 am) upto (not including) Aug 2 at midnight (0:00 am).

    lstash count program:haproxy --from "Aug 1" --to "Aug 2"

Grep all haproxy log messages using for one day (Aug 24 1 0:00 am upto and including Aug 2 23:59).

    lstash grep program:haproxy --from "Aug 1" --to "Aug 2"

Assuming today is Sep 1 2014. Count all haproxy log messages in the previous month.

    lstash count program:haproxy --anchor yesterday --from firstday --to today -d
    time range: [2014-08-01 00:00:00 +0200..2014-09-01 00:00:00 +0200]
    logstash-2014.07.31: 1
    logstash-2014.08.01: 13
    logstash-2014.08.02: 14
    logstash-2014.08.03: 1654
    logstash-2014.08.04: 6
    logstash-2014.08.05: 20
    logstash-2014.08.06: 219
    logstash-2014.08.07: 32
    logstash-2014.08.08: 14
    logstash-2014.08.09: 28
    logstash-2014.08.10: 799
    logstash-2014.08.11: 18
    logstash-2014.08.12: 8
    logstash-2014.08.13: 23
    logstash-2014.08.14: 25
    logstash-2014.08.15: 69
    logstash-2014.08.16: 19
    logstash-2014.08.17: 1160
    logstash-2014.08.18: 284
    logstash-2014.08.19: 61
    logstash-2014.08.20: 26
    logstash-2014.08.21: 16
    logstash-2014.08.22: 145
    logstash-2014.08.23: 72
    logstash-2014.08.24: 792
    logstash-2014.08.25: 31
    logstash-2014.08.26: 33
    logstash-2014.08.27: 51
    logstash-2014.08.28: 8
    logstash-2014.08.29: 23
    logstash-2014.08.30: 25
    logstash-2014.08.31: 69
    5633

## Using lstash as a gem in your project

Add this line to your application's Gemfile:

    gem 'lstash'

And then execute:

    bundle

Usage:

    bundle console

Connect to elasticsearch and create the Lstash client

    elasticsearch = Elasticsearch::Client.new(url: ENV['ES_URL'])
    client = Lstash::Client.new(elasticsearch, debug: true)

Create the query

    query = Lstash::Query.new('program:haproxy', from: 'today', to: 'now')

Count example

    client.count(query)

Grep example

    client.grep(query) do |message|
      puts message
    end

## Publishing the gem to RubyGems.org

1. Build the gem

        $ gem build lstash
        Successfully built RubyGem
        Name: lstash
        Version: 0.2.0
        File: lstash-0.2.0.gem

2. Pushing your gem to RubyGems.org

        $ gem push lstash-0.2.0.gem
        Pushing gem to RubyGems.org...
        Successfully registered gem: lstash (0.2.0)

See [RubyGems.org documention](https://guides.rubygems.org/) for more info.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## Release 1.0.0

* BREAKING CHANGE: default options changed
  * Default --from today changed to --from yesterday
  * Default --to now changed to --to today
  * This ensures that by default lstash counts or greps in yesterdays logging.
* Fixed bug which caused empty range (e.g. --from today --to today) to incorrectly return non-zero count and logging.
* Upgrade elasticsearch gem from version ~> 0.4 to ~> 7.17.7.
* Update queries and field selectors to be compatible with Elasticsearch version 7.
* Increase scroll step size for grep from 2 minutes to 1 hour (current Elasticsearch can handle it).
* Add --wildcard / --no-wildcard option to use logstash-* wildcard instead of iterating over indices directly.
  * For the count command --wildcard is faster so that's the default for count.
  * For the grep command --no-wildcard is faster so that's the default for grep.
* Dockerize development and add GitHub action for testing.
* Moved repo from kdgm/lstash to kdgm/lstash.
* Rubocop fixes

## Release 0.2.0

Merge branch 'feature/fix/hashie-warnings' into develop

* [fix] pin faraday to a compatible version
* [enh] upgrade to ruby 2.4(.6)
* [fix] suppress Hashie warnings; it would generate a warning for each log line (on stdout) leading to very large output

### 0.1.4 / 2015-05-29

Enhancements

* Change default time steps from 1 hour to 120 seconds to aid in grepping from datasets with many events per minute.

### 0.1.3 / 2015-05-28

Enhancements

* Remove troublesome development dependencies (autotest).

### 0.1.2 / 2015-01-02

Enhancements

* Double Elasticsearch::Client request timeout to 120 seconds to prevent Timeout::Error (Faraday::TimeoutError).

### 0.1.1 / 2014-10-13

Enhancements

* Refactor to iterate over hour periods instead of full indices. This improves the performance a lot.
* Increase timeout to prevent Faraday::TimeoutError exceptions.

### 0.0.9 / 2014-09-22

Enhancements

* We only every print the 'message' field. To optimize we only query for the 'message' field
  to be returned in the results. This should cut down the size of the returned json documents
  which could potentially be (very) large.

### 0.0.8 / 2014-09-01

Bug Fixes

* Require missing 'uri' (instead of using typhoed or patron HTTP transport clients).

### 0.0.7 / 2014-09-01

Bug Fixes

* Use Typhoed transport instead of Patron. The C-extension of Patron doesn't play nice with JRuby.

### 0.0.6 / 2014-09-01

Bug Fixes

* Pushing version 0.0.5 failed and it was yanked from rubygems.org. Need to push new version 0.0.6.

### 0.0.5 / 2014-09-01

Bug Fixes

* Running as binary didn't load HTTP client properly. Add 'patron' dependency
  to force loading an appropriate HTTP client.

### 0.0.4 / 2014-08-29

Bug Fixes

* Use .ruby-[version|gemset] to support any ruby environment manager.

### 0.0.3 / 2014-08-28

Bug Fixes

* Run CI on travis-ci.org.
* Fixate timezone to assumed timezone for specs.
* Run Ruby 2.1.1 instead of 2.1.0.

### 0.0.2 / 2014-08-28

Enhancements

* Updated documentation.
* Rename debug option (-v) to (-d).

### 0.0.1 / 2014-08-28

Initial release.


spec/lstash_spec.rb
require "spec_helper"
require "lstash/cli"

describe Lstash do
  it "should have a version number" do
    expect(Lstash::VERSION).not_to be nil
  end
end


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "lstash"

require "rspec/its"

require "timecop"

ENV["ES_URL"] = nil
ENV["TZ"] = "Europe/Amsterdam" # Test in a specific timezone.

RSpec.configure do |config|
  config.order = "random"

  # Suggestions taken from http://railscasts.com/episodes/413-fast-tests
  #
  # Focus on specific specs by tagging with `:focus`
  # or use `fit` instead of `it`.
  # ```ruby
  # it "focus test", :focus do
  # end
  #
  # fit "focus test" do
  # end
  # ```
  config.alias_example_to :fit, focus: true
  config.filter_run focus: true
  config.run_all_when_everything_filtered = true
end

require "stringio"

def capture_stdout
  old = $stdout
  $stdout = fake = StringIO.new
  yield
  fake.string
ensure
  $stdout = old
end

def capture_stderr
  old = $stderr
  $stderr = fake = StringIO.new
  yield
  fake.string
ensure
  $stderr = old
end


spec/lstash/cli_spec.rb
require "spec_helper"
require "lstash/cli"

class Lstash::CLI < Lstash::CLIBase
  def self.exit_on_failure?
    false
  end
end

describe Lstash::CLI do
  before(:all) do
    Timecop.freeze("2014-08-01 14:58")
  end
  after(:all) do
    Timecop.return
  end

  context "options" do
    subject { Lstash::CLI.options(args) }

    let(:args) { %w[extract --time from --to to --one --two --three --four] }

    its(:keys) { should eq args }
  end

  context "count" do
    context "with full URI" do
      let(:args) { %w[count "*" --es-url http://localhost:9200] }
      it "should succeed" do
        client = double("client")

        allow(Lstash::Client).to receive(:new).and_return(client)
        allow(client).to receive(:count).and_return(1000)

        expect {
          output = capture_stdout { Lstash::CLI.start(args) }
          expect(output).to eq "1000\n"
        }.not_to raise_error
      end
    end

    context "with valid arguments" do
      let(:args) { %w[count "program:haproxy" --es-url localhost] }

      it "should succeed" do
        client = double("client")

        allow(Lstash::Client).to receive(:new).and_return(client)
        allow(client).to receive(:count).and_return(100)

        output = capture_stdout { Lstash::CLI.start(args) }

        expect(output).to eq "100\n"
      end
    end

    context "with invalid --es-url" do
      let(:args) { %w[count "program:haproxy" --es-url ''] }

      it "should print error message" do
        output = capture_stderr { Lstash::CLI.start(args) }

        expect(output).to include("Failed to open TCP connection to '':9200")
      end
    end

    context "without query" do
      let(:args) { %w[] }
      it "should print help message" do
        output = capture_stdout { Lstash::CLI.start(args) }

        expect(output).to match("Commands:\n  rspec count QUERY")
      end
    end

    context "with anchor date" do
      let(:args) { %w[count program:haproxy --from firstday --to today --anchor yesterday --no-wildcard] }

      it "should return correct count" do
        es_client = double("es_client")

        allow(Elasticsearch::Client).to receive(:new) { es_client }
        allow_any_instance_of(Lstash::Client).to receive(:validate_shards!).and_return(true)

        expect(es_client).to receive(:count).exactly(31 * 24).times.and_return("count" => 100)

        output = capture_stdout { Lstash::CLI.start(args) }
        expect(output).to match((31 * 24 * 100).to_s)
      end
    end

    context "without anchor date" do
      let(:args) { %w[count program:haproxy --from yesterday --to today --no-wildcard] }

      it "should return correct count" do
        es_client = double("es_client")

        allow(Elasticsearch::Client).to receive(:new) { es_client }
        allow_any_instance_of(Lstash::Client).to receive(:validate_shards!).and_return(true)

        expect(es_client).to receive(:count).exactly(24).times.and_return("count" => 100)

        output = capture_stdout { Lstash::CLI.start(args) }
        expect(output).to match((24 * 100).to_s)
      end
    end
  end
end


lib/lstash/query.rb
require "time"
require "date"
require "ostruct"

module Lstash
  class Query
    class FormatError < StandardError; end

    class QueryMissing < StandardError; end

    LOGSTASH_PREFIX = "logstash-".freeze
    WILDCARD_QUERY = "*".freeze

    attr_accessor :from, :to

    def initialize(query_string = nil, arguments = {})
      @query_string = query_string

      @anchor = time_parse(arguments[:anchor], "today")
      @from = time_parse(arguments[:from], "yesterday")
      @to = time_parse(arguments[:to], "today")

      @to = Time.now if @to > Time.now # prevent accessing non-existing times / indices
    end

    def index_name(date)
      "#{LOGSTASH_PREFIX}#{date.strftime("%Y.%m.%d")}"
    end

    def wildcard_indices
      "logstash-*"
    end

    def search(from, size)
      {
        sort: sort_order,
        _source: %w[message],
        query: filter,
        from: from,
        size: size
      }
    end

    def filter
      {
        bool: {
          must: es_query,
          filter: es_filter
        }
      }
    end

    def each_period(step, &block)
      # iterate over the whole range in blocks of the specified period
      time_iterate(@from.utc, @to.utc - 1, step) do |start_at|
        yield index_name(start_at.to_date),
              Query.new(@query_string,
                anchor: @anchor,
                from: start_at,
                to: start_at + step)
      end
    end

    private

    def time_iterate(start_time, end_time, step, &block)
      while start_time < end_time
        yield(start_time)
        start_time += step
      end
    end

    def time_parse(time_or_string, default)
      return time_or_string if time_or_string.is_a? Time
      time_string = begin
        time_or_string.strip
      rescue
        nil
      end
      time_string ||= default
      case time_string
      when "firstday"
        midnight_at_beginning_of_month
      when "now"
        Time.now
      when "today"
        midnight_today
      when "yesterday"
        midnight_yesterday
      else
        Time.parse(time_string)
      end
    rescue ArgumentError
      raise FormatError, "Invalid time format: #{time_string}"
    end

    def query_string
      q = begin
        @query_string.dup.strip
      rescue
        ""
      end
      q = WILDCARD_QUERY if q.empty?
      q
    end

    def sort_order
      # return results in order of ascending timestamp
      [{"@timestamp" => {order: "asc"}}]
    end

    def es_query
      [
        {
          query_string: {
            query: query_string
          }
        }
      ]
    end

    def es_filter
      {
        range: {
          "@timestamp" => {
            gte: to_msec(from),
            lt: to_msec(to)
          }
        }
      }
    end

    # Return the date of the first day of date's month
    def midnight_at_beginning_of_month
      Date.new(anchor_time.year, anchor_time.month, 1).to_time
    end

    def midnight_today
      Date.today.to_time
    end

    def midnight_yesterday
      (Date.today - 1).to_time
    end

    def anchor_time
      @anchor
    end

    def to_msec(time)
      time.to_i * 1000
    end
  end
end


lib/lstash/client.rb
require "logger"
require "date"

class NullLogger < Logger
  def initialize(*args)
  end

  def add(*args, &block)
  end
end

module Lstash
  class Client
    class ConnectionError < StandardError; end

    class ShardMismatchError < StandardError; end

    PER_PAGE = 5000 # best time, lowest resource usage
    COUNT_STEP = 3600 # 1 hours
    GREP_STEP = 3600 # 1 hour

    def initialize(es_client, options = {})
      raise ConnectionError, "No elasticsearch client specified" if es_client.nil?

      @es_client = es_client
      @logger = options[:logger] || (options[:debug] ? debug_logger : NullLogger.new)
      @wildcard = options[:wildcard]
    end

    def count(query)
      @logger.debug "count from=#{query.from} to=#{query.to}"

      count = 0
      use_wildcard = @wildcard.nil? ? true : @wildcard
      if use_wildcard
        count = count_messages(query.wildcard_indices, query)
      else
        count = 0
        query.each_period(COUNT_STEP) do |index, hour_query|
          count += count_messages(index, hour_query)
        end
      end
      @logger.debug "total count=#{count}"
      count
    end

    def grep(query)
      @logger.debug "grep from=#{query.from} to=#{query.to}"

      count = 0
      use_wildcard = @wildcard.nil? ? false : @wildcard
      if use_wildcard
        grep_messages(query.wildcard_indices, query) do |message|
          count += 1
          yield message if block_given?
        end
      else
        query.each_period(GREP_STEP) do |index, hour_query|
          grep_messages(index, hour_query) do |message|
            count += 1
            yield message if block_given?
          end
        end
      end

      @logger.debug "total count=#{count}"
      count
    end

    private

    def count_messages(index, query)
      result = @es_client.count(index: index, body: {query: query.filter})
      validate_shards!(result["_shards"])
      @logger.debug "count index=#{index} from=#{query.from.utc} to=#{query.to.utc} count=#{result["count"]}"
      result["count"]
    end

    def grep_messages(index, query)
      messages = nil
      scroll_params = {}
      offset = 0
      method = :search
      while messages.nil? || messages.count > 0
        result = @es_client.send(method, {
          index: index,
          scroll: "5m",
          body: (method == :search) ? query.search(offset, PER_PAGE) : scroll_params
        })

        validate_shards!(result["_shards"])

        messages = result["hits"]["hits"]
        offset += messages.count
        scroll_params = {scroll_id: result["_scroll_id"]}

        messages.each do |h|
          next if h["_source"].nil?
          yield h["_source"]["message"] if block_given?
        end

        method = :scroll

        @logger.debug "grep index=#{index} from=#{query.from.utc} to=#{query.to.utc} count=#{offset}"
      end
      @es_client.clear_scroll(body: scroll_params) unless scroll_params.empty?
    end

    def debug_logger
      logger = Logger.new($stderr)
      logger.formatter = proc do |severity, datetime, progname, msg|
        "#{datetime} #{msg}\n"
      end
      logger
    end

    def validate_shards!(shards)
      if shards["total"] != shards["successful"]
        raise ShardMismatchError, "Shard mismatch: total: #{shards["total"]}, successful: #{shards["successful"]}"
      end
    end
  end
end


lib/lstash/cli.rb
# external dependencies
require "thor"
require "uri"
require "elasticsearch"

# local files we need
require "lstash/query"
require "lstash/client"
require "lstash/version"

module Lstash
  TRANSPORT_REQUEST_TIMEOUT = 120 # 2 minute request timeout

  class CLIBase < Thor
    class << self
      def shared_options
        method_option :anchor, banner: "YYYY-mm-dd", aliases: "-a", desc: "The 'firstday' is relative to this anchor date", default: "today"
        method_option :from, banner: "YYYY-mm-dd [HH:MM:SS]", aliases: "-f", desc: "Start date/time, 'now', 'today', 'yesterday', or 'firstday'", default: "yesterday"
        method_option :to, banner: "YYYY-mm-dd [HH:MM:SS]", aliases: "-t", desc: "End date/time, 'now', 'today', 'yesterday', or 'firstday'", default: "today"
        method_option :es_url, banner: "http://localhost:9200", aliases: "-e", desc: "Elasticsearch URL or set ES_URL environment variable"
        method_option :debug, desc: "Log debugging info to stderr", aliases: "-d", type: :boolean, default: false
        method_option :wildcard, desc: "Use index wildcard to query all logstash-* indices (fast for count, slow for grep)", type: :boolean
      end
    end
  end

  class CLI < CLIBase
    long_desc <<-LONGDESC
      Grep log messages matching the QUERY from Logstash in ascending timestamp order
      and output to stdout. QUERY can use Apache Lucene query parser syntax.

      Example to grep HAProxy log messages from the beginning of this month upto now

        lstash grep 'program:haproxy' --from firstday --to now
    LONGDESC
    desc "grep QUERY", "Grep log messages from Logstash"
    shared_options
    def grep(query_string)
      run_command(query_string) do |es_client, query|
        Lstash::Client.new(es_client, options).grep(query) do |message|
          puts message
        end
      end
    end

    long_desc <<-LONGDESC
      Count log messages matching the QUERY from Logstash and output this count to stdout.
      QUERY can use Apache Lucene query parser syntax.

      Example to count the number of HAProxy log messages in yesterdays month.

        lstash count 'program:haproxy' --from firstday --to today --anchor yesterday
    LONGDESC
    desc "count QUERY", "Count number of log messages matching the QUERY"
    shared_options
    def count(query_string)
      run_command(query_string) do |es_client, query|
        count = Lstash::Client.new(es_client, options).count(query)
        puts count
      end
    end

    long_desc "Print the lstash version"
    desc "version", "print lstash version"
    def version
      puts Lstash::VERSION
    end

    private

    def run_command(query_string)
      es_client = ::Elasticsearch::Client.new(
        url: options[:es_url] || ENV["ES_URL"] || "http://localhost:9200",
        log: ENV["DEBUG"] == "true",
        transport_options: {request: {timeout: TRANSPORT_REQUEST_TIMEOUT}}
      )
      query = Lstash::Query.new(query_string, options)

      yield es_client, query
    rescue => e
      options[:debug] ? raise(e) : raise(Thor::Error.new(e.message))
    end

    # Make sure we exit on failure with an error code
    class << self
      protected

      def exit_on_failure?
        true
      end
    end
  end
end
